src/main/kotlin/com/exactpro/th2/conn/dirty/fix/brokenconn/strategy/StrategyState.kt

```diff
@@ -34,46 +33,79 @@ class StrategyState(val config: RuleConfiguration? = null) {
     val startTime: Instant = Instant.now()
     val type = config?.ruleType ?: RuleType.DEFAULT
     val batchMessageCache: CompositeByteBuf = Unpooled.compositeBuffer()
-    val messageIDs = Collections.synchronizedList(ArrayList<MessageID>())
-
+    val messageIDs: MutableList<MessageID> = Collections.synchronizedList(ArrayList<MessageID>())
 
     private val writeLock = ReentrantLock()
-    private val missedMessagesCache: MutableMap<Int, ByteBuf> = ConcurrentHashMap<Int, ByteBuf>()
-
+    private val missedMessagesCache: MutableMap<Long, ByteBuf> = ConcurrentHashMap<Long, ByteBuf>()
     private val batchMessageCacheSize = AtomicInteger(0)
-    fun getBatchMessageCacheSize() = batchMessageCacheSize.get()
 
     private val missedIncomingMessagesCount = AtomicInteger(0)
-    fun getMissedIncomingMessagesCount() = missedIncomingMessagesCount.get()
-
-    private val missedOutgoingMessagesCount = AtomicInteger(0)
-    fun getMissedOutgoingMessagesCount() = missedOutgoingMessagesCount.get()
-
-    val transformedIncomingMessagesCount = AtomicInteger(0)
+    fun updateMissedIncomingMessagesCountIfCondition(condition: (Int) -> Boolean): Boolean = writeLock.withLock {
+        var updated = false
+        missedIncomingMessagesCount.updateAndGet {
+            if(condition(it + 1)) {
+                updated = true
+                it + 1
+            } else {
+                it
+            }
+        }
+        updated
+    }
 
-    fun incrementMissedIncomingMessages() { missedIncomingMessagesCount.incrementAndGet() }
-    fun incrementIncomingTransformedMessages() { transformedIncomingMessagesCount.incrementAndGet() }
+    private val transformedIncomingMessagesCount = AtomicInteger(0)
+    fun getTransformedIncomingMessagesCount() = writeLock.withLock { transformedIncomingMessagesCount.get() }
+    fun transformIfCondition(condition: (Int) -> Boolean, transform: () -> Unit): Boolean = writeLock.withLock {
+        var updated = false
+        transformedIncomingMessagesCount.updateAndGet {
+            if(condition(it + 1)) {
+                updated = true
+                transform()
+                it + 1
+            } else {
+                it
+            }
+        }
+        updated
+    }
 
-    fun addMissedMessageToCache(sequence: Int, message: ByteBuf) {
+    private val missedOutgoingMessagesCount = AtomicInteger(0)
+    fun addMissedMessageToCacheIfCondition(sequence: Long, message: ByteBuf, condition: (Int) -> Boolean): Boolean {
         writeLock.withLock {
-            missedMessagesCache[sequence] = message
-            missedOutgoingMessagesCount.incrementAndGet()
+            var updated = false
+            missedOutgoingMessagesCount.updateAndGet {
+                if(condition(it + 1)) {
+                    missedMessagesCache[sequence] = message
+                    updated = true
+                    it + 1
+                } else {
+                    it
+                }
+            }
```

Comment:
If we are already under the lock maybe we don't need the atomic here?..
Also, if I remember correctly, the lambda in `udate*` methods on atomic can be executed more than one time (if we have concurrent writes). So, we must be careful with what we try to execute there